### PR TITLE
Feature/aruco speedup (rebased with 4.x)

### PR DESCRIPTION
**Pull request by AleksandrPanov**
Added improvements to the ArUco module made by the author @urbste. New PR was created to fix issues and rebase with 4.x (and to avoid regression in original PR).

- Original PR: opencv#2790
- [Original article](https://www.researchgate.net/publication/325787310_Speeded_Up_Detection_of_Squared_Fiducial_Markers)

This PR:
- adds new 3 performance tests (19 configurations)
- enables old tests with distance parameter 0.4 for ArUco3 (CV_ArucoBoardPose test)
- fix several "index" bugs
- removes global threshold parameters - to add this parameters need to add PImpl pattern and create `ArucoDetector` class (like QRCodeDetector class). I suggest return global threshold parameters in next PR after create `ArucoDetector` class and API changes (this changes could significantly increase "git diff").

Configuration - Ryzen 7 2700x, 32GB 2933 mhz DDR4
Performance tests result (image size is 1440x1440 - the number of pixels is the same as in FHD image with size 1920x1080):

- with global threshold parameters:

| Algorithm | Parameters | Number of markers | Result, ms |
|--|--|--|--|
| base alg | default | 1   | mean=6.93   median=6.97   min=4.61   stddev=1.20 (17.3%)|
| arcuo3 | tau_c=32, tau_i=0.15 | 1   | mean=1.52   median=1.48   min=1.41   stddev=0.10 (6.5%) |
| arcuo3 | tau_c=16, tau_i=0.008 |  1  | mean=1.46   median=1.46   min=1.37   stddev=0.04 (3.0%) |
| base alg | default | 9   | mean=8.49   median=8.55   min=5.96   stddev=1.35 (15.9%)|
| arcuo3 | tau_c=32, tau_i=0.15 | 9   | mean=2.03   median=2.03   min=1.67   stddev=0.10 (4.7%) |
| arcuo3 | tau_c=16, tau_i=0.008 |  9  | mean=2.03   median=2.02   min=1.91   stddev=0.07 (3.4%) |
| base alg | default | 100 | mean=17.91  median=18.05  min=13   stddev=1.78 (10.0%)|
| arcuo3 | tau_c=32, tau_i=0.15 | 100 | mean=5.09   median=5.22   min=4.53   stddev=0.32 (6.3%) |
| arcuo3 | tau_c=16, tau_i=0.008 |  100| mean=5.11   median=5.13   min=4.67   stddev=0.29 (5.7%) |

- without global threshold parameters (these parameters will be returned in the next PR):

| Algorithm | Parameters | Number of markers | Result, ms |
|--|--|--|--|
| base alg | default | 1   | mean=6.84   median=6.69   min=4.7   stddev=0.96 (14.0%)|
| arcuo3 | tau_c=32, tau_i=0.15 | 1   | mean=3.03   median=3.08   min=2.51   stddev=0.20 (6.5%) |
| arcuo3 | tau_c=16, tau_i=0.008 |  1  | mean=2.94   median=3.03   min=2.31   stddev=0.29 (9.9%) |
| base alg | default | 9   | mean=8.39   median=8.02   min=6.08   stddev=1.44 (17.1%)|
| arcuo3 | tau_c=32, tau_i=0.15 | 9   | mean=3.80   median=3.80   min=2.98   stddev=0.32 (8.4%) |
| arcuo3 | tau_c=16, tau_i=0.008 |  9  | mean=3.65   median=3.73   min=3.09   stddev=0.28 (7.8%) |
| base alg | default | 100 | mean=18.40  median=19.29  min=14.09   stddev=1.95 (10.6%)|
| arcuo3 | tau_c=32, tau_i=0.15 | 100 | mean=8.23   median=8.14   min=8.04   stddev=0.22 (2.6%) |
| arcuo3 | tau_c=16, tau_i=0.008 |  100| mean=8.46   median=8.20   min=8.08   stddev=0.45 (5.4%) |

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/aruco/src/aruco.cpp [hunk 1]
```diff
@@ -679,16 +766,30 @@ static void _identifyCandidates(InputArray _image, vector< vector< vector< Point
     vector< int > rotated(ncandidates, 0);
     vector< uint8_t > validCandidates(ncandidates, 0);
 
+    const int min_perimeter = params->minSideLengthCanonicalImg * params->minSideLengthCanonicalImg;
```
Comment (AleksandrPanov) [inline modules/aruco/src/aruco.cpp:769]: "perimeter" is calculated by multiplying the sides of the square

Comment (AleksandrPanov) [reply modules/aruco/src/aruco.cpp]: this was bug, fixed

File: modules/aruco/perf/perf_aruco.cpp [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+            pitch = (pitch + 70) % 360;
+        }
+        imwrite("tile_test" + std::to_string(iter) + ".jpg", tileImage);
+        return std::make_pair(std::move(tileImage), std::move(idCorners));
+    }
+};
+
+TEST(EstimateAruco, ArucoThird)
+{
+    setNumThreads(1);
```
Comment (asmorkalov) [inline modules/aruco/perf/perf_aruco.cpp:214]: It's hack, is not it?

**Comment by urbste** [comment]
@AleksandrPanov : great work! Thanks a lot for finishing this PR. I just did not find the time. :/

File: modules/aruco/test/test_boarddetection.cpp [hunk 3]
```diff
@@ -181,8 +185,13 @@ void CV_ArucoBoardPose::run(int) {
     cameraMatrix.at< double >(1, 2) = imgSize.height / 2;
     Mat distCoeffs(5, 1, CV_64FC1, Scalar::all(0));
 
+    double max_dist = 0.4;
+    // aruco3 detection is a bit worse from large distances it seems
+    if (run_with == checkWithParameter::USE_ARUCO3) {
+        max_dist = 0.2;
+    }
```
Comment (AleksandrPanov) [inline modules/aruco/test/test_boarddetection.cpp:192]: In fact, a too small perimeter marker is created (marker perimeter < 32 * 4). Need set minSideLengthCanonicalImg as 16.

File: modules/aruco/src/aruco.cpp [hunk 4]
```diff
@@ -439,25 +453,19 @@ static void _detectInitialCandidates(const Mat &grey, vector< vector< Point2f >
...
-static void _detectCandidates(InputArray _image, vector< vector< vector< Point2f > > >& candidatesSetOut,
+static void _detectCandidates(InputArray _grayImage, vector< vector< vector< Point2f > > >& candidatesSetOut,
                               vector< vector< vector< Point > > >& contoursSetOut, const Ptr<DetectorParameters> &_params) {
 
-    Mat image = _image.getMat();
-    CV_Assert(image.total() != 0);
-
-    /// 1. CONVERT TO GRAY
-    Mat grey;
-    _convertToGrey(image, grey);
```
Comment (AleksandrPanov) [inline modules/aruco/src/aruco.cpp:450]: input image was gray already

File: modules/aruco/src/aruco.cpp [hunk 5]
```diff
@@ -667,14 +704,8 @@ static void _identifyCandidates(InputArray _image, vector< vector< vector< Point
     int ncandidates = (int)_candidatesSet[0].size();
     vector< vector< Point2f > > accepted;
     vector< vector< Point2f > > rejected;
-
     vector< vector< Point > > contours;
 
-    CV_Assert(_image.getMat().total() != 0);
-
-    Mat grey;
-    _convertToGrey(_image.getMat(), grey);
```
Comment (AleksandrPanov) [inline modules/aruco/src/aruco.cpp:676]: input image was gray already

File: modules/aruco/src/aruco.cpp [hunk 6]
```diff
@@ -667,14 +704,8 @@ static void _identifyCandidates(InputArray _image, vector< vector< vector< Point
     int ncandidates = (int)_candidatesSet[0].size();
     vector< vector< Point2f > > accepted;
     vector< vector< Point2f > > rejected;
-
     vector< vector< Point > > contours;
 
-    CV_Assert(_image.getMat().total() != 0);
```
Comment (AleksandrPanov) [inline modules/aruco/src/aruco.cpp:673]: this condition was checked already (in detectMarkers)

File: modules/aruco/src/aruco.cpp [hunk 7]
```diff
@@ -568,17 +576,21 @@ static int _getBorderErrors(const Mat &bits, int markerSize, int borderSize) {
...
 static uint8_t _identifyOneCandidate(const Ptr<Dictionary>& dictionary, InputArray _image,
-                                  vector<Point2f>& _corners, int& idx,
-                                  const Ptr<DetectorParameters>& params, int& rotation)
+                                  const vector<Point2f>& _corners, int& idx,
+                                  const Ptr<DetectorParameters>& params, int& rotation,
+                                  const float scale = 1.f)
 {
-    CV_Assert(_corners.size() == 4);
-    CV_Assert(_image.getMat().total() != 0);
-    CV_Assert(params->markerBorderBits > 0);
```
Comment (AleksandrPanov) [inline modules/aruco/src/aruco.cpp:576]: these conditions were checked already (in detectMarkers)

File: modules/aruco/perf/perf_aruco.cpp [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+static inline double getMaxDistance(map<int, vector<Point2f> > &golds, const vector<int>& ids,
+                                               const vector<vector<Point2f> >& corners)
+{
+    double distance = 0.;
+    for (size_t i = 0; i < ids.size(); i++)
+    {
+        int id = ids[i];
+        const auto gold_corners = golds.find(id);
+        if (gold_corners != golds.end())
+        for (int c = 0; c < 4; c++)
```
Comment (asmorkalov) [inline modules/aruco/perf/perf_aruco.cpp:164]: Please fix indentation. Also it case of id is missing in golds the function returns 0 and test passes silently. I propose to return some very large value to highlight the issue.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

File: modules/aruco/perf/perf_aruco.cpp [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+        detectorParams->minSideLengthCanonicalImg = 32;
+        detectorParams->minMarkerLengthRatioOriginalImg = 0.04f / numMarkersInRow;
+    }
+    MarkerPainter painter(markerSize);
+    auto image_map = painter.getProjectMarkersTile(numMarkersInRow, detectorParams, dictionary);
+
+    // detect markers
+    vector<vector<Point2f> > corners;
+    vector<int> ids;
+    TEST_CYCLE_N(10)
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/perf/perf_aruco.cpp]: removed, insert just `TEST_CYCLE()`

File: modules/aruco/samples/detector_params.yml [hunk 10]
```diff
@@ -21,3 +21,10 @@ perspectiveRemoveIgnoredMarginPerCell: 0.13
 maxErroneousBitsInBorderRate: 0.04
 minOtsuStdDev: 5.0
 errorCorrectionRate: 0.6
+
+# new aruco functionality
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/samples/detector_params.yml]: fixed

File: modules/aruco/src/aruco.cpp [hunk 11]
```diff
@@ -137,6 +143,10 @@ bool DetectorParameters::readDetectorParameters(const FileNode& fn, Ptr<Detector
     checkRead |= readParameter(fn["maxErroneousBitsInBorderRate"], params->maxErroneousBitsInBorderRate);
     checkRead |= readParameter(fn["minOtsuStdDev"], params->minOtsuStdDev);
     checkRead |= readParameter(fn["errorCorrectionRate"], params->errorCorrectionRate);
+    // new aruco functionality
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/src/aruco.cpp]: fixed

File: modules/aruco/perf/perf_aruco.cpp (hunk 8 above)
Comment (AleksandrPanov) [reply modules/aruco/perf/perf_aruco.cpp]: inserted `distMap` with `std::numeric_limits<double>::max()` default distance. Now `getMaxDistance` return max value of `distMap`. And if `ids` has no gold id, then `getMaxDistance` will return `std::numeric_limits<double>::max()`.

**Review by asmorkalov — APPROVED** [review]

File: modules/aruco/src/aruco.cpp [hunk 12]
```diff
@@ -137,6 +143,10 @@ bool DetectorParameters::readDetectorParameters(const FileNode& fn, Ptr<Detector
     checkRead |= readParameter(fn["maxErroneousBitsInBorderRate"], params->maxErroneousBitsInBorderRate);
     checkRead |= readParameter(fn["minOtsuStdDev"], params->minOtsuStdDev);
     checkRead |= readParameter(fn["errorCorrectionRate"], params->errorCorrectionRate);
+    // new aruco 3 functionality
+    checkRead |= readParameter(fn["useAruco3Detection"], params->useAruco3Detection);
+    checkRead |= readParameter(fn["minSideLengthCanonicalImg"], params->minSideLengthCanonicalImg);
+    checkRead |= readParameter(fn["minMarkerLengthRatioOriginalImg"], params->minMarkerLengthRatioOriginalImg);
     return checkRead;
```
Comment (alalek) [inline modules/aruco/src/aruco.cpp:150]: `false` result for old .xml files (from existed opencv versions) is an regression.

File: modules/aruco/src/aruco.cpp (hunk 7 above)
Comment (alalek) [reply modules/aruco/src/aruco.cpp:576]: It makes sense to keep them as `CV_DbgAssert()` anyway.

There are several static code analyzers which don't care about callers code but needs some checks.

File: modules/aruco/src/aruco.cpp (hunk 4 above)
Comment (alalek) [reply modules/aruco/src/aruco.cpp:450]: Add checks about assumptions of this code:
- not `.empty()`
- `CV_CheckType()`

File: modules/aruco/src/aruco.cpp (hunk 12 above)
Comment (AleksandrPanov) [reply modules/aruco/src/aruco.cpp:150]: There is no `false` result for old .xml files, because ` |=` operator is using.

Comment (AleksandrPanov) [reply modules/aruco/src/aruco.cpp:150]: Also, test `can_find_diamondmarkers` with read old .xml files was added in prev PR. This test successfully completed.  

Only `&=` operator could add `false` result for old .xml files.

File: modules/aruco/src/aruco.cpp [hunk 13]
```diff
@@ -655,10 +667,35 @@ static void correctCornerPosition( vector< Point2f >& _candidate, int rotate){
...
+static size_t _findOptPyrImageForCanonicalImg(
+        const std::vector<Mat>& img_pyr,
+        const int scaled_width,
+        const int cur_perimeter,
+        const int min_perimeter) {
+
+    size_t optLevel = 0;
+    float dist = std::numeric_limits<float>::max();
+    for (size_t i = 0; i < img_pyr.size(); ++i) {
+        const float scale = img_pyr[i].cols / static_cast<float>(scaled_width);
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/src/aruco.cpp]: `scaled_width` checked after resize image.
added `CV_Assert(scaled_width > 0);`

File: modules/aruco/src/aruco.cpp (hunk 7 above)
Comment (AleksandrPanov) [reply modules/aruco/src/aruco.cpp:576]: added `CV_DbgAssert`()

File: modules/aruco/src/aruco.cpp (hunk 4 above)
Comment (AleksandrPanov) [reply modules/aruco/src/aruco.cpp:450]: these conditions were checked already (in detectMarkers)

added `CV_DbgAssert`

**Comment by AleksandrPanov** [comment]
@alalek thx for review, I fixed your remarks

File: modules/aruco/src/aruco.cpp (hunk 6 above)
Comment (alalek) [reply modules/aruco/src/aruco.cpp:673]: assumption checks are necessary, see above

File: modules/aruco/test/test_boarddetection.cpp [hunk 14]
```diff
@@ -168,7 +172,7 @@ class CV_ArucoBoardPose : public cvtest::BaseTest {
 CV_ArucoBoardPose::CV_ArucoBoardPose() {}
 
 
-void CV_ArucoBoardPose::run(int) {
+void CV_ArucoBoardPose::run(int run_with) {
```
Comment (alalek) [inline modules/aruco/test/test_boarddetection.cpp:175]: Wrong usage. Don't change semantic of parameters.

Check the base class:

```
    // the main procedure of the test
    virtual void run( int start_from );
```

---

Use derived class or template.

Comment (AleksandrPanov) [reply modules/aruco/test/test_boarddetection.cpp]: removed `run_with` and add ArucoAlgParams to test classes constructor. Also field `Ptr<aruco::DetectorParameters> params` was added to test classes.

Comment (AleksandrPanov) [reply modules/aruco/test/test_boarddetection.cpp]: @alalek, fixed

File: modules/aruco/src/aruco.cpp (hunk 6 above)
Comment (AleksandrPanov) [reply modules/aruco/src/aruco.cpp:673]: fixed

File: modules/aruco/test/test_arucodetection.cpp [hunk 15]
```diff
@@ -534,6 +541,11 @@ TEST(CV_AprilTagDetectionPerspective, algorithmic) {
     test.safe_run(CV_ArucoDetectionPerspective::USE_APRILTAG);
 }
 
+TEST(CV_Aruco3DetectionPerspective, algorithmic) {
+    CV_Aruco3DetectionPerspective test;
+    test.safe_run(CV_ArucoDetectionPerspective::USE_ARUCO3);
```
Comment (alalek) [inline modules/aruco/test/test_arucodetection.cpp:546]: The issue with parameter misusing with this test too.

https://github.com/opencv/opencv/blob/4.5.5/modules/ts/include/opencv2/ts.hpp#L407-L411

Comment (AleksandrPanov) [reply modules/aruco/test/test_arucodetection.cpp]: Fixed:
parameter `tryWith` was removed, parameter `ArucoAlgParams` was added as field of class `CV_ArucoDetectionPerspective`

**Comment by zamlz** [comment]
@AleksandrPanov I have some questions regarding the changes in this PR.

I'm currently using the Aruco detection algorithm in an older version of OpenCV. I've heard from other sources that the opencv implementation was outdated because of the GPL licensing issues #2242. But @asmorkalov recently commented that a new independent implementation was brought into OpenCV with this PR.

From my limited experience here, it seems like the primary improvement here was improving the processing speed (I also briefly skimmed the attached paper), but speed is not a huge concern for my use case at the moment. I'm curious about the following things however.

- Is there a noticeable improvement to accuracy from these changes as well?
- Could you elaborate on what _"fix several 'index' bugs"_ means here? Would these bugs potentially effect accuracy depending on how I'm using Aruco?

**Pull request merged**